ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/BeaconStateMutators.java

```diff
@@ -98,18 +110,85 @@ public void initiateValidatorExit(MutableBeaconState state, int index) {
         .set(
             index,
             validator
-                .withExitEpoch(exitQueueEpoch)
+                .withExitEpoch(validatorExitContext.exitQueueEpoch)
                 .withWithdrawableEpoch(
-                    exitQueueEpoch.plus(specConfig.getMinValidatorWithdrawabilityDelay())));
+                    validatorExitContext.exitQueueEpoch.plus(
+                        specConfig.getMinValidatorWithdrawabilityDelay())));
+  }
+
+  /**
+   * This function implements an optimized version of exitQueueEpoch and exitQueueChurn calculation,
+   * compared to the `initiateValidatorExit` reference implementation.
+   *
+   * @param state
+   * @return
+   */
+  public Supplier<ValidatorExitContext> createValidatorExitContextSupplier(
+      final BeaconState state) {
+    return Suppliers.memoize(
+        () -> {
+          final ValidatorExitContext validatorExitContext =
+              new ValidatorExitContext(beaconStateAccessors.getValidatorChurnLimit(state));
+
+          validatorExitContext.exitQueueEpoch = UInt64.ZERO;
+
+          final List<Validator> exitedValidatorsInExitQueueEpoch = new ArrayList<>();
+
+          for (Validator validator : state.getValidators()) {
+            final UInt64 validatorExitEpoch = validator.getExitEpoch();
+            if (validatorExitEpoch.equals(FAR_FUTURE_EPOCH)) {
```

Comment:
How about this? it could be clearer I think
```java
            if (validatorExitEpoch.equals(validatorExitContext.exitQueueEpoch)) {
              exitedValidatorsInExitQueueEpoch.add(validator);
            } else if (!validatorExitEpoch.equals(FAR_FUTURE_EPOCH)
                && validatorExitEpoch.isGreaterThan(
                    validatorExitContext.exitQueueEpoch)) { // New biggest exit epoch found
              validatorExitContext.exitQueueEpoch = validatorExitEpoch;
              exitedValidatorsInExitQueueEpoch.clear();
              exitedValidatorsInExitQueueEpoch.add(validator);
            }
```

Reply:
My brain considers that version less clear :)

Reply:
Yeh i think where the code is already at is more clear...

Reply:
arguably, 
```
    for (Validator validator : state.getValidators()) {
      final UInt64 validatorExitEpoch = validator.getExitEpoch();
      if (!validatorExitEpoch.equals(FAR_FUTURE_EPOCH)
          && validatorExitEpoch.isGreaterThanOrEqualTo(validatorExitContext.exitQueueEpoch)) {
        if (validatorExitEpoch.isGreaterThan(validatorExitContext.exitQueueEpoch)) {
          exitedValidatorsInExitQueueEpoch.clear();
          validatorExitContext.exitQueueEpoch = validatorExitEpoch;
        }
        exitedValidatorsInExitQueueEpoch.add(validator);
      }
    }
```
clearest i could come up with...

Reply:
hopefully you like that because I committed it :)

Reply:
@rolfyone  Honestly, I don't! :) my brain still prefers my version, which has no nested ifs and clear separated conditions. And in this version we are doing an additional comparison.